src/passes/Asyncify.cpp

```diff
@@ -841,6 +856,55 @@ struct AsyncifyFlow : public Pass {
     // don't want it to be seen by asyncify itself.
     return builder->makeCall(ASYNCIFY_GET_CALL_INDEX, {}, none);
   }
+
+  // Given a function that is not instrumented - because we proved it doesn't
+  // need it, or depending on the whitelist/blacklist - add assertions that
+  // verify that property at runtime.
+  // Note that it is ok to run code while sleeping (if you are careful not
+  // to break assumptions in the program!) - so what is actually
+  // checked here is if the state *changes* in an uninstrumented function.
+  // That is, if in an uninstrumented function, a sleep should not begin
+  // from any call.
+  void addAssertsInNonInstrumented(Function* func) {
+    auto oldState = builder->addVar(func, i32);
+    // Add a check at the function entry.
+    func->body = builder->makeSequence(
+      builder->makeLocalSet(oldState,
+                            builder->makeGlobalGet(ASYNCIFY_STATE, i32)),
+      func->body);
+    // Add a check around every call.
+    struct Walker : PostWalker<Walker> {
+      void visitCall(Call* curr) { handleCall(curr); }
+      void visitCallIndirect(CallIndirect* curr) { handleCall(curr); }
+      void handleCall(Expression* call) {
```

Comment:
I know Asyncify doesn't work with tail calls at all yet, but we could future-proof this by skipping the assert after a tail call, since it will never be reached.

Reply:
Thanks, good point. I added asserts here, as I think we will need to think about another kind of assertion for that case, so the asserts will force us to remember.